### PR TITLE
[Memory] Better Memory Utilization for PyHealth

### LINKED ISSUE #332: Improve memory usage of `BaseDataset.load_data()`

### PR DESCRIPTION
**Pull request by Logiquo**
This should be able to close #332 .

### CODE REVIEW

**Review by jhnwu3 — CHANGES_REQUESTED** [review]
Hey do you know which dask version you used? Probably would be good to add this to the pyprojectoml. 

Will look deeper into the CI and other implementation details later, have to go wander Japan for a bit 😂

File: pyproject.toml [hunk 1]
```diff
@@ -39,6 +39,9 @@ dependencies = [
   "pandas~=2.3.1",
   "pandarallel~=1.6.5",
   "pydantic~=2.11.7",
+  "dask[complete]~=2025.11.0",
+  "litdata~=0.2.58",
+  "xxhash~=3.6.0",
```
Comment (Logiquo) [inline pyproject.toml:44]: New dependencies are specified here.

Comment (jhnwu3) [reply pyproject.toml:44]: ah im blind.

File: pyhealth/data/data.py [hunk 2]
```diff
@@ -20,9 +20,9 @@ class Event:
 
     event_type: str
     timestamp: datetime
-    attr_dict: Mapping[str, any] = field(default_factory=dict)
+    attr_dict: Mapping[str, Any] = field(default_factory=dict)
```
Comment (Logiquo) [inline pyhealth/data/data.py:23]: Small type hint fix here, `any` is a function for iterables, and `Any` is the correct type hint.

File: pyhealth/data/data.py [hunk 3]
```diff
@@ -119,65 +119,76 @@ class Patient:
...
     Attributes:
         patient_id (str): Unique patient identifier.
-        data_source (pl.DataFrame): DataFrame containing all events, sorted by timestamp.
-        event_type_partitions (Dict[str, pl.DataFrame]): Dictionary mapping event types to their respective DataFrame partitions.
+        data_source (pd.DataFrame): DataFrame containing all events, sorted by timestamp.
+        event_type_partitions (Dict[str, pd.DataFrame]): Dictionary mapping event types to their respective DataFrame partitions.
     """
 
-    def __init__(self, patient_id: str, data_source: pl.DataFrame) -> None:
+    def __init__(self, patient_id: str, data_source: pd.DataFrame) -> None:
```
Comment (Logiquo) [inline pyhealth/data/data.py:126]: The polars will spawn processes. Given this will be ran in litdata.optimize in a multi-processes environment, nested multi-process will cause a hang. Thus, we can this to pandas.

File: pyhealth/data/data.py [hunk 4]
```diff
@@ -119,65 +119,76 @@ class Patient:
...
             return df
 
+    def get_events_py(self, **kawargs) -> List[Event]:
+        """Type-safe wrapper for get_events."""
+        return self.get_events(**kawargs, return_df=False) # type: ignore
+
+    def get_events_df(self, **kawargs) -> pd.DataFrame:
+        """DataFrame wrapper for get_events."""
+        return self.get_events(**kawargs, return_df=True) # type: ignore
+
```
Comment (Logiquo) [inline pyhealth/data/data.py:183]: Added for type hint purpose.

File: pyhealth/tasks/mortality_prediction_stagenet_mimic4.py [hunk 5]
```diff
@@ -185,46 +187,51 @@ def __call__(self, patient: Any) -> List[Dict[str, Any]]:
                 all_icd_times.append(time_from_previous)
 
             # Get lab events for this admission
-            labevents_df = patient.get_events(
+            labevents_df = patient.get_events_df(
                 event_type="labevents",
                 start=admission_time,
                 end=admission_dischtime,
-                return_df=True,
             )
```
Comment (Logiquo) [inline pyhealth/tasks/mortality_prediction_stagenet_mimic4.py:194]: It returns a pandas dataframe here, any task that uses `get_event_df` (or `get_event(..., return_df=True)`) will need to update their code to be compitable with pandas.

File: pyhealth/tasks/base_task.py [hunk 6]
```diff
@@ -1,15 +1,15 @@
...
+import dask.dataframe as dd
 
 
 class BaseTask(ABC):
     task_name: str
     input_schema: Dict[str, Union[str, Type]]
     output_schema: Dict[str, Union[str, Type]]
 
-    def pre_filter(self, df: pl.LazyFrame) -> pl.LazyFrame:
+    def pre_filter(self, df: dd.DataFrame) -> dd.DataFrame:
```
Comment (Logiquo) [inline pyhealth/tasks/base_task.py]: It becomes dask dataframe here, any task override this need to update.

File: pyhealth/datasets/utils.py [hunk 7]
```diff
@@ -331,10 +332,10 @@ def get_dataloader(
     Returns:
         A DataLoader instance for the dataset.
     """
+    dataset.set_shuffle(shuffle)
     dataloader = DataLoader(
         dataset,
         batch_size=batch_size,
-        shuffle=shuffle,
         collate_fn=collate_fn_dict_with_padding,
     )
```
Comment (Logiquo) [inline pyhealth/datasets/utils.py:341]: python native dataloader does not support shuffle on IterableDataset, it must be shuffled at dataset level.

File: pyhealth/datasets/splitter.py [hunk 8]
```diff
@@ -40,9 +36,9 @@ def split_by_visit(
         int(len(dataset) * ratios[0]) : int(len(dataset) * (ratios[0] + ratios[1]))
     ]
     test_index = index[int(len(dataset) * (ratios[0] + ratios[1])) :]
-    train_dataset = torch.utils.data.Subset(dataset, train_index)
-    val_dataset = torch.utils.data.Subset(dataset, val_index)
-    test_dataset = torch.utils.data.Subset(dataset, test_index)
+    train_dataset = SampleSubset(dataset, train_index) # type: ignore
+    val_dataset = SampleSubset(dataset, val_index) # type: ignore
+    test_dataset = SampleSubset(dataset, test_index) # type: ignore
```
Comment (Logiquo) [inline pyhealth/datasets/splitter.py:41]: torch native Subset does not support iterable dataset, we use a custom-defined iterable dataset here.

File: pyhealth/datasets/sample_dataset.py [hunk 9]
```diff
@@ -141,45 +152,153 @@ def build(self) -> None:
...
-        Returns:
-            int: The number of samples.
+class SampleSubset(IterableDataset):
+    """A subset of the SampleDataset.
+
+    Args:
+        sample_dataset (SampleDataset): The original SampleDataset.
+        indices (List[int]): List of indices to include in the subset.
+    """
+
```
Comment (Logiquo) [inline pyhealth/datasets/sample_dataset.py:205]: Add new SampleSubset to support creating subset of a SampleDataset for train/val/test split.

File: pyhealth/datasets/sample_dataset.py [hunk 10]
```diff
@@ -141,45 +152,153 @@ def build(self) -> None:
...
+        base_dataset.set_shuffle(False) # Disable shuffling when creating subset
+        self.dataset, self._length = self._build_subset_dataset(
+            base_dataset, indices
+        )
+
+    def _build_subset_dataset(
+        self, base_dataset: StreamingDataset, indices: Sequence[int]
+    ) -> Tuple[StreamingDataset, int]:
+        """Create a StreamingDataset restricted to the provided indices."""
+        if len(base_dataset.subsampled_files) != len(base_dataset.region_of_interest):
```
Comment (Logiquo) [inline pyhealth/datasets/sample_dataset.py:220]: Modified based on official litdata train_test_split https://github.com/Lightning-AI/litData/blob/main/src/litdata/utilities/train_test_split.py

File: pyhealth/datasets/mimic4.py [hunk 11]
```diff
@@ -210,19 +211,26 @@ def __init__(
         note_config_path: Optional[str] = None,
         cxr_config_path: Optional[str] = None,
         dataset_name: str = "mimic4",
+        cache_dir: str | Path | None = None,
+        num_workers: int = 1,
+        mem_per_worker: str | int = "8GB",
+        compute: bool = True,
```
Comment (Logiquo) [inline pyhealth/datasets/mimic4.py:217]: Add other dataset will need to add these new args.

File: pyhealth/datasets/mimic4.py [hunk 12]
```diff
@@ -210,19 +211,26 @@ def __init__(
...
+        super().__init__(
+            root=f"{str(ehr_root)},{str(note_root)},{str(cxr_root)}",
+            tables=(ehr_tables or []) + (note_tables or []) + (cxr_tables or []),
+            dataset_name=dataset_name,
+            cache_dir=cache_dir,
+            num_workers=num_workers,
+            mem_per_worker=mem_per_worker,
+            compute=False, # defer compute to later, we need to aggregate all sub-datasets first
+            dev=dev,
+        )
```
Comment (Logiquo) [inline pyhealth/datasets/mimic4.py:229]: important to assign root, tables, dataset_name, dev to calculate a correct default cache path.

File: pyhealth/datasets/mimic4.py [hunk 13]
```diff
@@ -260,21 +276,18 @@ def __init__(
...
-        self._collected_global_event_df = None
-        self._unique_patient_ids = None
-
+        if compute:
+            _ = self._joined_cache()
         log_memory_usage("Completed MIMIC4Dataset init")
 
-    def _combine_data(self) -> pl.LazyFrame:
+    def load_data(self) -> dd.DataFrame:
         """
```
Comment (Logiquo) [inline pyhealth/datasets/mimic4.py:291]: one should override `load_data` if it requires custom logic, since there is no `.global_event_df` available. The existance of a dask dataframe in class field would cause a multi-process failure, causing it unable to use multiple worker to process the sample.

File: pyhealth/datasets/base_dataset.py [hunk 14]
```diff
@@ -56,42 +71,55 @@ def path_exists(path: str) -> bool:
...
+    (bucket_id, merged_cache, task) = input
+    path = f"{merged_cache}/bucket={bucket_id}"
+    # This is more efficient than reading patient by patient
+    grouped = pd.read_parquet(path).groupby("patient_id")
+
+    for patient_id, patient_df in grouped:
+        patient = Patient(patient_id=str(patient_id), data_source=patient_df)
+        for sample in task(patient):
+            # Schema is too complex to be handled by LitData, so we pickle the sample here
+            yield _pickle(sample)
```
Comment (Logiquo) [inline pyhealth/datasets/base_dataset.py:128]: It's important to define these functions (outside of a class) here to avoid issue for multi-processes environment.

File: pyhealth/datasets/base_dataset.py [hunk 15]
```diff
@@ -110,96 +138,233 @@ def __init__(
...
+                "dataset_name": self.dataset_name,
+                "dev": self.dev,
+            },
+            sort_keys=True,
+        )
+        cache_dir = Path(platformdirs.user_cache_dir(appname="pyhealth")) / str(
+            uuid.uuid5(uuid.NAMESPACE_DNS, id_str)
+        )
+        print(f"No cache_dir provided. Using default cache dir: {cache_dir}")
+        self._cache_dir = cache_dir
```
Comment (Logiquo) [inline pyhealth/datasets/base_dataset.py:215]: Compute default cache dir. I think this should be unique enough?

File: pyhealth/datasets/base_dataset.py [hunk 16]
```diff
@@ -110,96 +138,233 @@ def __init__(
...
+                        estimated_mem_usage // partition_size
+                    ) + 1  # Calculate number of partitions based on memory usage
+                    n_partitions = max(
+                        n_partitions, self.num_workers
+                    )  # At least num_workers partitions
+
+                    bucket = global_event_df["patient_id"].apply(
+                        lambda pid: _patient_bucket(pid, n_partitions),
+                        meta=("patient_id", "int"),
+                    )
```
Comment (Logiquo) [inline pyhealth/datasets/base_dataset.py:328]: Split dataframe into bucket based on paitent id, enable faster processing downstream.

File: pyhealth/datasets/base_dataset.py [hunk 17]
```diff
@@ -320,30 +482,39 @@ def get_patient(self, patient_id: str) -> Patient:
...
+        with open(f"{path}/index.json", "rb") as f:
+            n_partitions = json.load(f)["n_partitions"]
+        bucket = _patient_bucket(patient_id, n_partitions)
+        path = f"{path}/bucket={bucket}"
+        df = pd.read_parquet(path)
+        patient = Patient(
+            patient_id=patient_id,
+            data_source=df[df["patient_id"] == patient_id],
+        )
+        return patient
```
Comment (Logiquo) [inline pyhealth/datasets/base_dataset.py:502]: Read from relevant bucket only, this would be much faster for larger dataset.

**Comment by Logiquo** [comment]
Closed in favour of #687
